facia-press/app/frontpress/FapiFrontPress.scala

```diff
@@ -489,13 +453,62 @@ trait FapiFrontPress extends EmailFrontPress with Logging {
 
     contentApiResponse.map(Option(_)).fallbackTo(Future.successful(None))
   }
-
 }
 
-object CapiUrl {
-  def extractId(urlString: String): String = {
-    val uri = new URI(urlString)
-    val path = uri.getPath()
-    path.stripPrefix("/")
+object Enrichment extends Logging {
+  def enrichSnap(embedUri: Option[String], beforeEnrichment: EnrichedContent, collection: Collection, wsClient: WSClient)
+                (implicit executionContext: ExecutionContext): Future[EnrichedContent] = {
+    embedUri match {
+      case Some(embedUri) =>
+        wsClient.url(embedUri).get().map { response =>
+          Json.fromJson[EmbedJsonHtml](response.json) match {
+            case JsSuccess(embed, _) =>
+              beforeEnrichment.copy(embedHtml = Some(embed.html))
+            case _ =>
+              log.warn(s"An embed had invalid json format, and won't be pressed. $embedUri for collection ${collection.id}")
+              beforeEnrichment
+          }
+        } recover {
+          case _ => {
+            log.warn(s"A request to an embed uri failed, embed won't be pressed. $embedUri for collection ${collection.id}")
+            beforeEnrichment
+          }
+        }
+      case _ => Future(beforeEnrichment)
+    }
+  }
+
+  def enrichInteractive(atomId: Option[String], beforeEnrichment: EnrichedContent, collection: Collection, capiClient: CapiContentApiClient)
+                       (implicit executionContext: ExecutionContext): Future[EnrichedContent] = {
+    atomId match {
+      case Some(atomId) => {
+        capiClient.getResponse(ItemQuery(atomId)).map { response =>
+          response.interactive.flatMap { interactive =>
+            interactive.data match {
+              case atom: com.gu.contentatom.thrift.AtomData.Interactive =>
+                Some(beforeEnrichment.copy(embedHtml = Some(atom.interactive.html), embedCss = Some(atom.interactive.css), embedJs = atom.interactive.mainJS))
+              case _ =>
+                None
+            }
+          }.getOrElse({
+            handleInteractiveFailure(s"${atomId} for collection ${collection.id}")
+            beforeEnrichment
+          })
+        } recover { case _ => {
+          handleInteractiveFailure(s"${atomId} for collection ${collection.id}")
+          beforeEnrichment
```

Comment:
It seems a shame we've got three very similar failure paths in here. I wonder if there's a way to rewrite this to to flatten the error handling.